keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,522 @@
+want control over whose Job gets to run on the cluster. With a custom mutating
+webhook that adds `stopped: true` to all user-submitted Jobs, I can
+indefinitely delay Pod creation.
+
+### Notes/Constraints/Caveats
+
+System administrators may want to restrict who can set/modify the `stopped`
+field when creating or updating Jobs. This can be achieved with validating
+and/or mutating webhooks.
+
+When a Job is stopped, the time will be recorded as a part of the Job status.
+Users can infer how long a Job has been paused for with this field. This can be
+useful when making decisions around which Job should be resumed.
+
+A Job that is complete cannot be stopped.
+
+The feature will be gated behind a feature gate.
+
+### Risks and Mitigations
+
+1. Stopping an active Job will delete *all* active pods. Users must design
+   their application to gracefully handle this.
+
+## Design Details
+
+We propose adding a `stopped` field to the JobSpec API:
+
+```go
+type JobSpec struct {
+	// Stopped specifies whether the Job controller should create Pods or not. If
+	// a Job is created in the stopped state, no Pods are created by the Job
+	// controller. If a Job enters the stopped state (i.e. the flag goes from
+	// false to true), the Job controller will delete all Pods associated with
+	// this Job. This is an alpha field and requires enabling the StoppedJob
+	// feature gate. Defaults to false.
+	Stopped *bool `json:"stopped,omitempty"`
+
+	...
+}
+```
+
+As described in the comment, when the boolean is set to true, the
+controller-manager will abstain from creating Pods even if the Job's
+parallelism requirements aren't met. If the Job is already active and is
+updated with `stopped: true`, all Pods owned by this Job will be deleted.
+
+Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
+"Complete" and "Failed", we propose adding a new condition called "Stopped"
+that will be a part of the Job's status:
+
+```go
+type JobStatus struct {
+	// The latest available observations of an object's current state. Can be
+	// one of "Stopped", "Failed", or "Complete".
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	...
+}
+
+type JobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// JobStopped means the job has been stopped.
+	JobStopped JobConditionType = "Stopped"
+	// JobComplete means the job has completed its execution.
+	JobComplete JobConditionType = "Complete"
+	// JobFailed means the job has failed its execution.
+	JobFailed JobConditionType = "Failed"
+)
+```
+
+When a Job is stopped or created in the stopped state, a "Stopped" event will
+be recorded. Similarly, when a Job is resumed from its stopped state, a
+"Resumed" event will be recorded.
+
+In addition to the new JobConditionType, a new `StopTime` field is also
```

Comment:
I don't think we need this. The JobCondition has a timestamp: https://github.com/kubernetes/kubernetes/blob/a04b6e4b1671810ede5b8cacf4527741781d6fb9/staging/src/k8s.io/api/batch/v1/types.go#L188

Reply:
good catch, that should work! Do you think this will be obvious enough to users? I certainly didn't notice that that field can be used to infer stop time.

Reply:
Proper documentation should be enough. We don't want to introduce a new field just because a synonym one is not too obvious.